server/scrape_cache.go

```diff
@@ -0,0 +1,322 @@
+// scrapes.
+type scrapeCache struct {
+	iter uint64 // Current scrape iteration.
+
+	// How many series and metadata entries there were at the last success.
+	successfulCount int
+
+	// Parsed string to an entry with information about the actual label set
+	// and its storage reference.
+	series map[string]*cacheEntry
+
+	// Cache of dropped metric strings and their iteration. The iteration must
+	// be a pointer so we can update it.
+	droppedSeries map[string]*uint64
+
+	// seriesCur and seriesPrev store the labels of series that were seen
+	// in the current and previous scrape.
+	// We hold two maps and swap them out to save allocations.
+	seriesCur  map[uint64]labels.Labels
+	seriesPrev map[uint64]labels.Labels
+
+	metaMtx  sync.RWMutex
+	metadata map[string]*metaEntry
+}
+
+func newScrapeCache() *scrapeCache {
+	return &scrapeCache{
+		series:        map[string]*cacheEntry{},
+		droppedSeries: map[string]*uint64{},
+		seriesCur:     map[uint64]labels.Labels{},
+		seriesPrev:    map[uint64]labels.Labels{},
+		metadata:      map[string]*metaEntry{},
+	}
+}
+
+func (c *scrapeCache) iterDone(flushCache bool) {
+	c.metaMtx.Lock()
+	count := len(c.series) + len(c.droppedSeries) + len(c.metadata)
+	c.metaMtx.Unlock()
+
+	switch {
+	case flushCache:
+		c.successfulCount = count
+	case count > c.successfulCount*2+1000:
+		// If a target had varying labels in scrapes that ultimately failed,
+		// the caches would grow indefinitely. Force a flush when this happens.
+		// We use the heuristic that this is a doubling of the cache size
+		// since the last scrape, and allow an additional 1000 in case
+		// initial scrapes all fail.
+		flushCache = true
+	}
+
+	if flushCache {
+		// All caches may grow over time through series churn
+		// or multiple string representations of the same metric. Clean up entries
+		// that haven't appeared in the last scrape.
+		for s, e := range c.series {
+			if c.iter != e.lastIter {
+				delete(c.series, s)
+			}
+		}
+		for s, iter := range c.droppedSeries {
+			if c.iter != *iter {
+				delete(c.droppedSeries, s)
+			}
+		}
+		c.metaMtx.Lock()
+		for m, e := range c.metadata {
+			// Keep metadata around for 10 scrapes after its metric disappeared.
+			if c.iter-e.lastIter > 10 {
+				delete(c.metadata, m)
+			}
+		}
+		c.metaMtx.Unlock()
+
+		c.iter++
+	}
+
+	// Swap current and previous series.
+	c.seriesPrev, c.seriesCur = c.seriesCur, c.seriesPrev
+
+	// We have to delete every single key in the map.
+	for k := range c.seriesCur {
+		delete(c.seriesCur, k)
+	}
+}
+
+func (c *scrapeCache) get(met []byte) (*cacheEntry, bool) {
+	e, ok := c.series[string(met)]
+	if !ok {
+		return nil, false
+	}
+	e.lastIter = c.iter
+	return e, true
+}
+
+func (c *scrapeCache) addRef(met []byte, ref storage.SeriesRef, lset labels.Labels, hash uint64) {
+	if ref == 0 {
+		return
+	}
+	c.series[string(met)] = &cacheEntry{ref: ref, lastIter: c.iter, lset: lset, hash: hash}
```

Comment:
We are using a mutex to guard the `metadata` map but not this one and `droppedSeries`. Any reason?